### PR TITLE
Use config schema definition for validation.

### PR DESCRIPTION
**Pull request by tech4him1**
**- Summary**

Closes https://github.com/netlify/netlify-cms/issues/487.

I was just thinking about the best method for config validation, and this seemed like a viable option. [@Benaiah's validator](https://github.com/Benaiah/netlify-cms-config-examiner-rewrite) may be better, but I think it's worth a comparison.

This kind of basic format definitely does have the downside of only being able to statically check things. Most logic is available, though, and we can import constants from other files (valid formats, for example).

**- Test plan**

**- Description for the changelog**

Validate config with schema.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by verythorough** [comment]
Deploy preview for *netlify-cms-www* ready!

Built with commit 1093e7c05c7c84ce0f065bfca9eea089a214222e

https://deploy-preview-1363--netlify-cms-www.netlify.com

**Comment by verythorough** [comment]
Deploy preview for *cms-demo* ready!

Built with commit 1093e7c05c7c84ce0f065bfca9eea089a214222e

https://deploy-preview-1363--cms-demo.netlify.com

**Comment by tech4him1** [comment]
Example error messages. I'd like opinions on which of these are unclear and need updated.

- *`slug.encoding` should be equal to one of the allowed values*
  - Should we list the allowed values in the error?
- *`collections[0].fields` should contain a valid item*
  - This might be one where we need a specific error message -- it is in the case where a `title` field is missing.
- *'collections[0]' should be object*
  - This is in the case where a collection is not an object (`null`, a string, etc.).

**Comment by erquhart** [comment]
I think these are fine for a start, we can always improve moving forward.

**Comment by tech4him1** [comment]
@erquhart Can you review this on a "real" config file (e.g. Netlify, Smashing, etc.)?

**Comment by tech4him1** [comment]
@Benaiah Any thoughts on this?

**Comment by jake-101** [comment]
i can't wait for this!!

**Comment by tech4him1** [comment]
@erquhart @Benaiah @talves I think this should go in v2.0 if we are going for it -- that will avert any possible breaking changes caused by unintended behiviour that used to work.

File: src/constants/configSchema.js [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+                  items: {
+                    // ------- Each field: -------
+                    type: "object",
+                    properties: {
+                      name: { type: "string" },
+                      label: { type: "string" },
+                      widget: { type: "string" },
+                      required: { type: "boolean" },
+                    },
+                    required: ["name", "label", "widget"],
```
Comment (erquhart) [inline src/constants/configSchema.js]: `widget` isn't required, it's `string` by default.

File: src/constants/configSchema.js [hunk 2]
```diff
@@ -69,7 +69,7 @@ const getConfigSchema = () => ({
                       widget: { type: "string" },
                       required: { type: "boolean" },
                     },
-                    required: ["name", "label", "widget"],
+                    required: ["name", "label"],
```
Comment (tech4him1) [inline src/constants/configSchema.js]: Should we drop `label` too since that is a pending PR?

File: src/constants/configSchema.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                name: { type: "string" },
+                label: { type: "string" },
+                widget: { type: "string" },
+                required: { type: "boolean" },
+              },
+              required: ["name", "label"],
+            },
+          },
+        },
+        required: ["name", "label"],
```
Comment (tech4him1) [inline src/constants/configSchema.js]: Should we drop `label` too since that is a pending PR?

File: src/constants/configSchema.js [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                  items: {
+                    // ------- Each field: -------
+                    type: "object",
+                    properties: {
+                      name: { type: "string" },
+                      label: { type: "string" },
+                      widget: { type: "string" },
+                      required: { type: "boolean" },
+                    },
+                    required: ["name", "label"],
```
Comment (tech4him1) [inline src/constants/configSchema.js]: Should we drop `label` too since that is a pending PR?

Comment (erquhart) [reply src/constants/configSchema.js]: yeah, makes sense - I'm pushing another commit right now, will include that.

**Comment by erquhart** [comment]
CI will fail until I merge the new stuff from #1573.

**Pull request merged**